### PR TITLE
Thumbnails for the layer node

### LINKED ISSUE #1211: Display thumbnails in the node graph

### PR DESCRIPTION
**Pull request by 0HyperCube**
Closes #1211

Thumbnails in the node graph - building on TrueDoctor's node graph introspection mechanism.

ToDo:
- [x] Vector thumbnails
- [x] Image frame thumbnails
- [x] Group thumbnails

Styling as part of #1175

### CODE REVIEW

File: frontend/wasm/src/editor_api.rs [hunk 1]
```diff
@@ -122,10 +122,13 @@ impl JsEditorHandle {
...
 						js_sys::Float64Array::default()
 					};
-					updateImage(image.path, image.mime, &image.image_data, transform, document_id);
+					updateImage(image.path, image.node_id, image.mime, &image.image_data, transform, document_id);
 				}
 				#[cfg(feature = "tauri")]
-				fetchImage(image.path.clone(), image.mime, document_id, format!("http://localhost:3001/image/{:?}_{}", &image.path, document_id));
+				{
+					let identifier = format!("http://localhost:3001/image/{:?}_{}", &image.path, document_id);
+					fetchImage(image.path.clone(), image.node_id, image.mime, document_id, identifier);
```
Comment (nat-rix) [inline frontend/wasm/src/editor_api.rs:130]: some nitpicking:
- `.clone` isn't needed here, `image.path` can be consumed, it isn't needed later
- `&image.path`: the `&` isn't needed. All rust `format`-macros auto-ref their arguments, so they never consume later needed variables.

Comment (0HyperCube) [reply frontend/wasm/src/editor_api.rs:130]: Sure, this was just copied from here before.

File: node-graph/gcore/src/graphic_element/renderer/quad.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+impl core::ops::Mul<Quad> for DAffine2 {
+	type Output = Quad;
+
+	fn mul(self, rhs: Quad) -> Self::Output {
+		let mut output = Quad::default();
+		for (i, point) in rhs.0.iter().enumerate() {
+			output.0[i] = self.transform_point2(*point);
+		}
+		output
```
[the remark this answers is not in the document]
Comment (0HyperCube) [reply node-graph/gcore/src/graphic_element/renderer/quad.rs:48]: I didn't see that [arrays had map](https://github.com/rust-lang/rust/issues/75243).

File: node-graph/gcore/src/graphic_element/renderer.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		self.bounding_box_with_transform(self.transform * transform)
+	}
+}
+
+impl GraphicElementRendered for Artboard {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		self.graphic_group.render_svg(render, render_params)
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		let artboard_bounds = self.bounds.map(|[a, b]| (transform * Quad::from_box([a.as_dvec2(), b.as_dvec2()])).bounding_box());
```
Comment (nat-rix) [inline node-graph/gcore/src/graphic_element/renderer.rs:95]: Isn't that just the same as:
```rust
let artboard_bounds = self.bounds.map(|points| points.map(|c| transform.transform_point2(c.as_dvec2())))
```

File: node-graph/gcore/src/graphic_element/renderer.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+			.svg
+			.push(format!(r#"<image width="1" height="1" preserveAspectRatio="none" transform="matrix({transform})" href=""#).into());
+		let uuid = generate_uuid();
+		render.svg.push(SvgSegment::BlobUrl(uuid));
+		render.svg.push("\" />".into());
+		render.image_data.push((uuid, self.image.clone()))
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		let transform = self.transform * transform;
+		(transform.matrix2 != glam::DMat2::ZERO).then(|| (transform * Quad::from_box([DVec2::ZERO, DVec2::ONE])).bounding_box())
```
Comment (nat-rix) [inline node-graph/gcore/src/graphic_element/renderer.rs:113]: Similarily to line 95:
```rust
(transform.matrix2 != glam::DMat2::ZERO).then(|| [transform.translation, transform.transform_point2(DVec2::ONE)])

```

File: node-graph/gcore/src/graphic_element/renderer.rs (hunk 3 above)
Comment (0HyperCube) [reply node-graph/gcore/src/graphic_element/renderer.rs:95]: No, it isn't. Note that when rotating a rectangle, the axis aligned bounding box is no longer simply the transformed positions of the top right and bottom left corner:
image

File: node-graph/gcore/src/graphic_element/renderer.rs (hunk 4 above)
Comment (0HyperCube) [reply node-graph/gcore/src/graphic_element/renderer.rs:113]: No, it isn't. Note that when rotating a rectangle, the axis aligned bounding box is no longer simply the transformed positions of the top right and bottom left corner:
image

File: node-graph/gcore/src/graphic_element/renderer.rs (hunk 3 above)
Comment (nat-rix) [reply node-graph/gcore/src/graphic_element/renderer.rs:95]: oh right, somehow my brain ignored the existence of rotation.

File: node-graph/gcore/src/graphic_element/renderer.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+impl GraphicElementRendered for ImageFrame<Color> {
+	fn render_svg(&self, render: &mut SvgRender, _render_params: &RenderParams) {
+		let transform: String = format_transform_matrix(self.transform * render.transform);
+		render
+			.svg
+			.push(format!(r#"<image width="1" height="1" preserveAspectRatio="none" transform="matrix({transform})" href=""#).into());
+		let uuid = generate_uuid();
+		render.svg.push(SvgSegment::BlobUrl(uuid));
+		render.svg.push("\" />".into());
+		render.image_data.push((uuid, self.image.clone()))
```
Comment (Keavon) [inline node-graph/gcore/src/graphic_element/renderer.rs:109]: ChatGPT recommends this, is it applicable?

image

**Review by Keavon — APPROVED** [review]
Looks fantastic. Merging this now. Feel free to check out and respond to the one comment I left if it's relevant.

**Pull request merged**